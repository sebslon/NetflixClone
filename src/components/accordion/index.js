import React, { createContext, useState, useContext } from "react";

import {
  Body,
  Container,
  Header,
  Item,
  Inner,
  Title,
} from "./styles/accordion";

const ToggleContext = createContext();

export default function Accordion({ children, ...otherProps }) {
  return (
    <Container {...otherProps}>
      <Inner>{children}</Inner>
    </Container>
  );
}

Accordion.Title = function AccordionTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>;
};

Accordion.Item = function AccordionItem({ children, ...otherProps }) {
  const [toggleShow, setToggleShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ toggleShow, setToggleShow }}>
      <Item {...otherProps}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...otherProps }) {
  const { toggleShow, setToggleShow } = useContext(ToggleContext);

  return (
    <Header
      onClick={() => setToggleShow((prevState) => !prevState)}
      {...otherProps}
    >
      {children}
      {toggleShow ? (
        <img src="images/icons/close-slim.png" alt="close" />
      ) : (
        <img src="images/icons/add.png" alt="open" />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...otherProps }) {
  const { toggleShow } = useContext(ToggleContext);

  return toggleShow ? <Body {...otherProps}>{children}</Body> : null;
};
