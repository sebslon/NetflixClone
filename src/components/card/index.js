import { createContext, useContext, useState } from "react";

import {
  Container,
  Content,
  Entities,
  Feature,
  FeatureClose,
  FeatureTitle,
  FeatureText,
  Group,
  Item,
  Image,
  Maturity,
  Meta,
  SubTitle,
  Title,
  Text,
} from "./styles/card";

export const FeatureContext = createContext();

export default function Card({ children, ...otherProps }) {
  const [showFeature, setShowFeature] = useState(false);
  const [itemFeature, setItemFeature] = useState({});

  return (
    <FeatureContext.Provider
      value={{ showFeature, setShowFeature, itemFeature, setItemFeature }}
    >
      <Container {...otherProps}>{children}</Container>
    </FeatureContext.Provider>
  );
}

Card.Group = function CardGroup({ children, ...otherProps }) {
  return <Group {...otherProps}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>;
};

Card.SubTitle = function CardSubTitle({ children, ...otherProps }) {
  return <SubTitle {...otherProps}>{children}</SubTitle>;
};

Card.Text = function CardText({ children, ...otherProps }) {
  return <Text {...otherProps}>{children}</Text>;
};

Card.Entities = function CardEntity({ children, ...otherProps }) {
  return <Entities {...otherProps}>{children}</Entities>;
};

Card.Meta = function CardMeta({ children, ...otherProps }) {
  return <Meta {...otherProps}>{children}</Meta>;
};

Card.Feature = function CardFeature({ children, category, ...otherProps }) {
  const { showFeature, itemFeature, setShowFeature } = useContext(FeatureContext);

  return showFeature ? (
    <Feature {...otherProps}
      src={`/images/${category}/${itemFeature.genre}/${itemFeature.slug}/large.jpg`}
    >
      <Content>
        <FeatureTitle>{itemFeature.title}</FeatureTitle>
        <FeatureText>{itemFeature.description}</FeatureText>
        <FeatureClose onClick={() => setShowFeature(false)}>
          <img src="/images/icons/close.png" alt="close" />
        </FeatureClose>

        <Group margin="30px 0" flexDirection="row" alignItems="center">
          <Maturity rating={itemFeature.maturity}>
            {itemFeature.maturity > 12 ? "PG" : itemFeature.maturity}
          </Maturity>
          <FeatureText fontWeight="bold">
            {itemFeature.genre.charAt(0).toUpperCase() + itemFeature.genre.slice(1)}
          </FeatureText>
        </Group>
        {children}
      </Content>
    </Feature>
  ) : null;
};

Card.Item = function CardItem({ item, children, ...otherProps }) {
  const { setShowFeature, setItemFeature } = useContext(FeatureContext);

  return (
    <Item
      onClick={() => {
        setItemFeature(item);
        setShowFeature(true);
      }}
      {...otherProps}
    >
      {children}
    </Item>
  );
};

Card.Image = function CardImage({ ...otherProps }) {
  return <Image {...otherProps} />;
};
