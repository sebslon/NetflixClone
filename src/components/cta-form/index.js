import { Break, Container, Input, Button, Text } from './styles/cta-form';

export default function CtaForm({ children, ...otherProps }) {
  return (
    <Container {...otherProps}>{children}</Container>
  )
}

CtaForm.Input = function CtaFormInput({ ...otherProps }) {
  return <Input {...otherProps} />
}

CtaForm.Button = function CtaFormButton({ children, ...otherProps }) {
  return (
    <Button {...otherProps}>
      {children} <img src="images/icons/chevron-right.png" alt="Try now" />
    </Button>
  )
};

CtaForm.Text = function CtaFormText({ children, ...otherProps }) {
  return <Text {...otherProps}>{children}</Text>
}

CtaForm.Break = function CtaFormBreak({...otherProps}) {
  return <Break {...otherProps} />;
}