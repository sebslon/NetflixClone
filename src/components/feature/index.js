import { Container, Title, SubTitle } from './styles/feature';

export default function Feature({children, ...otherProps}) {
  return <Container {...otherProps}>{children}</Container>
};

Feature.Title = function FeatureTitle({children, ...otherProps}) {
  return <Title {...otherProps}>{children}</Title>
}

Feature.SubTitle = function FeatureSubTitle({children, ...otherProps}) {
  return <SubTitle {...otherProps}>{children}</SubTitle>
}