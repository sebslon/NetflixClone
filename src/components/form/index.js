import { Base, Container, Error, Title, Text, TextSmall, Link, Input, Submit } from './styles/form';

export default function Form({ children, ...otherProps }) {
  return <Container {...otherProps}>{children}</Container>
}

Form.Base = function FormBase({ children, ...otherProps }) {
  return <Base {...otherProps}>{children}</Base>
}

Form.Error = function FormError({ children, ...otherProps }) {
  return <Error {...otherProps}>{children}</Error>
}

Form.Title = function FormTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>
}

Form.Text = function FormText({ children, ...otherProps }) {
  return <Text {...otherProps}>{children}</Text>
}

Form.TextSmall = function FormTextSmall({ children, ...otherProps }) {
  return <TextSmall {...otherProps}>{children}</TextSmall>
}

Form.Link = function FormLink({ children, ...otherProps }) {
  return <Link {...otherProps}>{children}</Link>
}

Form.Input = function FormInput({ children, ...otherProps }) {
  return <Input {...otherProps}>{children}</Input>
}

Form.Submit = function FormSubmit({ children, ...otherProps }) {
  return <Submit {...otherProps}>{children}</Submit>
}