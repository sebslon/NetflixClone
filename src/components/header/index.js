import { Link } from 'react-router-dom';

import { Background, Container, ButtonLink, Logo } from './styles/header';

export default function Header({ background = true, children, ...otherProps }) {
  return background ? <Background {...otherProps}>{children}</Background> : children;
}

Header.Frame = function HeaderFrame({ children, ...otherProps }) {
  return <Container {...otherProps}>{children}</Container>;
}

Header.ButtonLink = function HeaderButtonLink({ children, ...otherProps }) {
  return <ButtonLink {...otherProps}>{children}</ButtonLink>;
}

Header.Logo = function HeaderLogo({ to, ...otherProps }) {
  return (
    <Link to={to}>
      <Logo {...otherProps} />
    </Link>
  )
}
