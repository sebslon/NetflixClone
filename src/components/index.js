export { default as Accordion } from './accordion'
export { default as CtaForm } from './cta-form'
export { default as Footer } from './footer';
export { default as Jumbotron } from './jumbotron';
