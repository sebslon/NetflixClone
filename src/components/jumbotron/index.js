import React from 'react';

import { Container, Inner, Item, Pane, Title, SubTitle, Image } from './styles/jumbotron';

export default function Jumbotron({ children, direction = 'row', ...otherProps }) {
  return (
    <Item>
      <Inner direction={direction}>
        {children}
      </Inner>
    </Item>
  )
}

Jumbotron.Container = function JumbotronContainer({ children, ...otherProps }) {
  return <Container {...otherProps}>{children}</Container>
}

Jumbotron.Pane = function JumbotronPane({ children, ...otherProps }) {
  return <Pane {...otherProps}>{children}</Pane>
}

Jumbotron.Title = function JumbotronTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>
}

Jumbotron.SubTitle = function JumbotronSubTitle({ children, ...otherProps }) {
  return <SubTitle {...otherProps}>{children}</SubTitle>
}

Jumbotron.Image = function JumbotronImage({ ...otherProps }) {
  return <Image {...otherProps} />
}