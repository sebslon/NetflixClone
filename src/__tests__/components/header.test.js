import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter as Router } from "react-router-dom";

import { Header } from "../../components";

describe("<Header />", () => {
  jest.mock("react-router-dom");

  it("Renders the basic <Header /> with a background", () => {
    const { container, getByText, getByTestId } = render(
      <Router>
        <Header>
          <Header.Frame>
            <Header.Logo src="/logo.svg" alt="Netflix" />
            <Header.TextLink active="true">TestLink</Header.TextLink>
          </Header.Frame>
        </Header>
      </Router>
    );

    expect(getByText("TestLink")).toBeTruthy();
    expect(getByTestId("header-bg")).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Renders the basic <Header /> without a background", () => {
    const { container, getByText, queryByTestId } = render(
      <Router>
        <Header background={false}>
          <Header.Frame>
            <Header.Logo src="/logo.svg" alt="Netflix" />
            <Header.ButtonLink>Sign In</Header.ButtonLink>
            <Header.TextLink active={false}>TestLink</Header.TextLink>
          </Header.Frame>
        </Header>
      </Router>
    );

    expect(getByText("TestLink")).toBeTruthy();
    expect(queryByTestId("header-bg")).toBeFalsy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Renders the full <Header /> with a background", () => {
    const { container, getByText, getByTestId } = render(
      <Router>
        <Header src="joker1" dontShowOnSmallViewPort>
          <Header.Frame>
            <Header.Group>
              <Header.Logo src="/images/logo.svg" alt="Netflix" />
              <Header.TextLink active={false} onClick={() => {}}>
                Series
              </Header.TextLink>
              <Header.TextLink active onClick={() => {}}>
                Films
              </Header.TextLink>
            </Header.Group>
            <Header.Group>
              <Header.Search searchTerm="Joker" setSearchTerm={() => {}} />
              <Header.Profile>
                <Header.Picture src="/images/test.png" />
                <Header.Dropdown>
                  <Header.Group>
                    <Header.Picture src="/images/test.png" />
                    <Header.TextLink>TestLink</Header.TextLink>
                  </Header.Group>
                  <Header.Group>
                    <Header.TextLink onClick={() => {}}>
                      Sign out
                    </Header.TextLink>
                  </Header.Group>
                </Header.Dropdown>
              </Header.Profile>
            </Header.Group>
          </Header.Frame>

          <Header.Feature>
            <Header.FeatureCallOut>Watch Joker Now</Header.FeatureCallOut>
            <Header.Text>Forever alone in a crowd...</Header.Text>
            <Header.PlayButton>Play</Header.PlayButton>
          </Header.Feature>
        </Header>
      </Router>
    );

    expect(getByTestId("search-input")).toBeTruthy();
    expect(getByTestId("search-input").value).toBe("Joker");
    userEvent.type(getByTestId("search-input"), {
      target: { value: "Simpsons" },
    });
    userEvent.click(getByTestId("search-click"));

    expect(getByText("Series")).toBeTruthy();
    expect(getByText("Films")).toBeTruthy();
    expect(getByText("TestLink")).toBeTruthy();
    expect(getByText("Watch Joker Now")).toBeTruthy();
    expect(getByText("Sign out")).toBeTruthy();
    expect(getByText("Forever alone in a crowd...")).toBeTruthy();
    expect(getByText("Play")).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
