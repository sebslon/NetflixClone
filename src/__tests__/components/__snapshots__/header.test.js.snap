// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Renders the basic <Header /> with a background 1`] = `
<div
  class="header__Background-djltcn-0 dlgEfJ"
  data-testid="header-bg"
>
  <div
    class="header__Container-djltcn-2 kjZpcb"
  >
    <a
      href="/"
    >
      <img
        alt="Netflix"
        class="header__Logo-djltcn-11 fdUOZV"
        src="/logo.svg"
      />
    </a>
    <p
      class="header__Link-djltcn-3 jKcjTD"
    >
      TestLink
    </p>
  </div>
</div>
`;

exports[`<Header /> Renders the basic <Header /> without a background 1`] = `
<div
  class="header__Container-djltcn-2 kjZpcb"
>
  <a
    href=""
  >
    <img
      alt="Netflix"
      class="header__Logo-djltcn-11 fdUOZV"
      src="/logo.svg"
    />
  </a>
  <a
    class="header__ButtonLink-djltcn-12 UaWRW"
    href=""
  >
    Sign In
  </a>
  <p
    class="header__Link-djltcn-3 dFxyeL"
  >
    TestLink
  </p>
</div>
`;

exports[`<Header /> Renders the full <Header /> with a background 1`] = `
<div
  class="header__Background-djltcn-0 nJJZH"
  data-testid="header-bg"
  src="joker1"
>
  <div
    class="header__Container-djltcn-2 kjZpcb"
  >
    <div
      class="header__Group-djltcn-1 NcEJn"
    >
      <a
        href=""
      >
        <img
          alt="Netflix"
          class="header__Logo-djltcn-11 fdUOZV"
          src="/images/logo.svg"
        />
      </a>
      <p
        class="header__Link-djltcn-3 dFxyeL"
      >
        Series
      </p>
      <p
        class="header__Link-djltcn-3 dFxyeL"
      >
        Films
      </p>
    </div>
    <div
      class="header__Group-djltcn-1 NcEJn"
    >
      <div
        class="header__Search-djltcn-7 grbakr"
      >
        <button
          class="header__SearchIcon-djltcn-8 jyMJiO"
          data-testid="search-click"
        >
          <img
            alt="Search"
            src="/images/icons/search.png"
          />
        </button>
        <input
          class="header__SearchInput-djltcn-9 kRDVcS"
          data-testid="search-input"
          placeholder="Search films and series"
          value="Joker"
        />
      </div>
      <div
        class="header__Profile-djltcn-10 fxBgPV"
      >
        <button
          class="header__Picture-djltcn-4 cKvker"
          src="/images/users//images/test.png.png"
        />
        <div
          class="header__Dropdown-djltcn-5 hPSwBU"
        >
          <div
            class="header__Group-djltcn-1 NcEJn"
          >
            <button
              class="header__Picture-djltcn-4 cKvker"
              src="/images/users//images/test.png.png"
            />
            <p
              class="header__Link-djltcn-3 dFxyeL"
            >
              TestLink
            </p>
          </div>
          <div
            class="header__Group-djltcn-1 NcEJn"
          >
            <p
              class="header__Link-djltcn-3 dFxyeL"
            >
              Sign out
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="header__Container-djltcn-2 header__Feature-djltcn-13 kjZpcb bZbAVy"
  >
    <p
      class="header__Text-djltcn-14 header__FeatureHeader-djltcn-15 kUavZC cnXqgA"
    >
      Watch Joker Now
    </p>
    <p
      class="header__Text-djltcn-14 kUavZC"
    >
      Forever alone in a crowd...
    </p>
    <button
      class="header__PlayButton-djltcn-6 jDMGzw"
    >
      Play
    </button>
  </div>
</div>
`;
